### DISCUSSION TITLE
Is client-side hashing with Nitrokey HSM 2 not possible?

### DISCUSSION THREAD

**Discussion by Vampire** [Keyfactor/signserver-ce#107] [Q&A]
We bought a Nitrokey HSM 2.
We put a SignServer instance in front of it.

Now I wanted to get signing with client-side hash computation working to minimize network traffic and work the HSM has to do.

With a PKCS#12 file and according worker this works fine.
But if I try to do the same with the PKCS#11 worker for the HSM and the `NONEwithRSA` algorithm, I get an error saying “No such algorithm: RSA/ECB/PKCS1Padding”.

Is this not supported using Nitrokey HSM 2?

**Comment by netmackan** [comment]
Could be that the HSM does not support that PKCS#11 mechanism (i.e. CKM_RSA_PKCS?) or that Java (SunPKCS11 no longer does?).
Maybe you could run Java with `-Djava.security.debug=sunpkcs11` and I believe at some point it would print out all supported mechanisms.

If you find the out something let us know.

See https://docs.oracle.com/en/java/javase/17/security/pkcs11-reference-guide1.html#JSSEC-GUID-7989F8B4-7260-4908-8203-99056B2D060E
